src/isotp_fast.c

```diff
@@ -990,11 +994,12 @@ int isotp_fast_recv(struct isotp_fast_ctx *ctx, struct can_filter sender, uint8_
 #endif /* CONFIG_ISOTP_FAST_BLOCKING_RECEIVE */
 
 int isotp_fast_send(struct isotp_fast_ctx *ctx, const uint8_t *data, size_t len,
-                    const isotp_fast_node_id target_addr, void *cb_arg)
+                    const uint8_t target_addr, uint8_t target_bus, void *cb_arg)
 {
-    const isotp_fast_can_id rx_can_id = (ctx->rx_can_id & 0xFFFF0000)
-                                        | (isotp_fast_get_target_addr(ctx->rx_can_id))
-                                        | (target_addr << ISOTP_FIXED_ADDR_TA_POS);
+    const uint32_t rx_can_id = (ctx->rx_can_id & 0xFF000000) | (target_bus << 20)
+                               | (isotp_fast_get_target_bus(ctx->rx_can_id) << 16)
+                               | (isotp_fast_get_target_addr(ctx->rx_can_id))
+                               | (target_addr << ISOTP_FIXED_ADDR_TA_POS);
```

Comment:
Similar to my comments elsewhere, I think we will struggle to get this merged in to Zephyr. In any event, the logic I originally built here would not work for, e.g. normal addressing, so some other approach will ultimately be required for that merge to be proposed anyway.

One option would be to make `rx_can_id` the function parameter. The simplest solution would then have the logic to decide on its value move outside the function, offloading the problem to the consumer of the API. That would be a bit of a shame.

We could improve on that by having an addressing mode, which would then behave differently depending on the mode selected. An optional mode, `ISOTP_FAST_ADDRESSING_MODE_CUSTOM`, would allow us to handle our use case.

Just sketching that out:

```c
enum isotp_fast_addressing_mode
{
    ISOTP_FAST_ADDRESSING_MODE_STANDARD,
    ISOTP_FAST_ADDRESSING_MODE_FIXED,
    ...
#ifdef CONFIG_ISOTP_FAST_ALLOW_CUSTOM_ADDRESSING
    ISOTP_FAST_ADDRESSING_MODE_CUSTOM
#endif
}

#ifdef CONFIG_ISOTP_FAST_ALLOW_CUSTOM_ADDRESSING
typedef uint32_t isotp_fast_get_rx_addr_custom_callback_t(struct isotp_fast_ctx *ctx, uint32_t target_addr);
#endif

struct isotp_fast_ctx
{
    ...
    enum isotp_fast_addressing_mode addressing_mode : 8;
#ifdef CONFIG_ISOTP_FAST_ALLOW_CUSTOM_ADDRESSING
    isotp_fast_get_rx_addr_custom_callback_t get_rx_addr;
#endif
    ...
}

uint32_t get_rx_can_id(struct isotp_fast_ctx *ctx, const uint32_t target_addr)
{
    switch (ctx->addressing_mode) {
        case ISOTP_FAST_ADDRESSING_MODE_STANDARD:
            return target_addr; // just return verbatim
        case ISOTP_FAST_ADDRESSING_MODE_FIXED:
            return (ctx->rx_can_id & 0xFFFF0000) | isotp_fast_get_target_addr(ctx->rx_can_id) | (target_addr & 0xFF);
#ifdef CONFIG_ISOTP_FAST_ALLOW_CUSTOM_ADDRESSING
        case ISOTP_FAST_ADDRESSING_MODE_CUSTOM:
            return ctx->get_rx_addr(ctx, target_addr);
#endif
    }
}
```

(I've just knocked the above up in the inline editor here, so it might be complete rubbish, but perhaps it gives a flavour of what I'm getting at.)

Reply:
Yep, that looks like a very good approach for upstreaming. We could even get rid of the Kconfigs for ID layout customization currently available in Zephyr upstream, as this could all be handled in a callback in the application.

Should we leave this for a future PR?

Reply:
> We could even get rid of the Kconfigs for ID layout customization

Yes, that would be a lot cleaner.

> Should we leave this for a future PR?

Yes. I might throw something together. I had been thinking it would be good to try to get this into an upstreamable state.